### PR TITLE
Update withState to accept function for initial state value

### PR DESCRIPTION
**Pull request by jjgillespie**

### CODE REVIEW

File: src/enhancers/withState.js [hunk 1]
```diff
@@ -1,18 +1,26 @@
...
   return {
     dependencies: {
       useState,
+      useMemo,
       [initialValueAlias]: initialValue,
+      [initialValueIsFunctionAlias]: Object.prototype.toString.call(initialValue) === '[object Function]',
     },
     initialize: `
-      const ${stateValue} = useState(${initialValueAlias});
+      const ${computedInitialValueAlias} = ${initialValueIsFunctionAlias}
```
Comment (jgretz) [inline src/enhancers/withState.js]: In thinking about the implementation for this, I'm leaning towards doing the check in the higher code and emitting either code set appropriately

File: src/enhancers/withState.js [hunk 2]
```diff
@@ -1,18 +1,26 @@
...
+  const initialValueIsFunctionAlias = generateNewVariable();
+  const computedInitialValueAlias = generateNewVariable();
   const stateValue = generateNewVariable();
 
   return {
     dependencies: {
       useState,
+      useMemo,
       [initialValueAlias]: initialValue,
+      [initialValueIsFunctionAlias]: Object.prototype.toString.call(initialValue) === '[object Function]',
```
Comment (jgretz) [inline src/enhancers/withState.js]: why not use typeof ?

**Review by jgretz — CHANGES_REQUESTED** [review]

File: src/enhancers/withState.js (hunk 2 above)
Comment (jjgillespie) [reply src/enhancers/withState.js]: I grabbed this implementation from lodash. This wasn't the article I read (http://es-discourse.com/t/why-typeof-is-no-longer-safe/15), but I got deep into MDN and a stack overflow discussion about some inconsistent behavior with typeof. It probably wouldn't happen in our limited case, but I felt safe with whatever lodash did...

File: src/enhancers/withState.js (hunk 1 above)
Comment (jjgillespie) [reply src/enhancers/withState.js]: Yeah that works too, I went back and forth with some different implementations. I'll update this a bit later

File: src/enhancers/withState.js (hunk 2 above)
Comment (jjgillespie) [reply src/enhancers/withState.js]: Okay I updated `withState` let me know what you think.

**Pull request merged**